File: Modules/EventSystem/src/EventManager.cpp
#include "EventManager.h"
#include <stdexcept>
#include <iostream>

EventManager* EventManager::instance = nullptr;

EventManager::EventManager(){

}

EventManager* EventManager::getInstance(){
    if(instance == nullptr){
        instance = new EventManager();
    }
    return instance;
}

void EventManager::subscribe(std::string name,AbstractEvent* event){
    std::cout << "Subscribe " << name << std::endl;
    try{
      
        std::vector<AbstractEvent*>& vec = mEvents.at(name);
        vec.push_back(event);
        std::cout << "Insert into "<< name << std::endl;
    }catch(std::out_of_range e){
      std::cout << name << " not registered. creating new one" << std::endl;
        std::vector<AbstractEvent*> vec;
        vec.push_back(event);
        mEvents.insert(std::pair<std::string,std::vector<AbstractEvent*>>(name,vec));
    }
}
void EventManager::fireEvent(std::string name,EventParam* param){
    try{
        std::vector<AbstractEvent*> vec = mEvents.at(name);
        //std::cout << "Fire Event: "<< name << " " << vec.size() << std::endl;
        for(auto it = vec.begin();it != vec.end();++it){
            (*it)->event(name,param);
        }
        delete param;
    } catch(std::out_of_range e){
      delete param;
      //throw e;
    }
}

File: Modules/rust-telemetry-fe/src/ts/EventManager.ts
class Event{

}

export class EventManager{
    private static instance?:EventManager = null; 
    private events:Map<string,Function[]> = new Map<string,Function[]>();
    private constructor(){}

    public static getInstance():EventManager{
        if(this.instance == null){
            this.instance = new EventManager();
        }
        return this.instance;
    }
    public subscribe(eventName:string,eventFunction:Function){
        if(!this.events.has(eventName)){
            this.events.set(eventName,[]);
            
        } 
        this.events.get(eventName).push(eventFunction);
    }
    public dispatch(eventName:string,param?:Object){
        if(this.events.has(eventName)){
            let eventFunctions = this.events.get(eventName)
            for(let i in eventFunctions){
                if(!eventFunctions.hasOwnProperty(i)) continue;
                eventFunctions[i](param);
            }
        }
    }
}